geoapps/io/validators.py

```diff
@@ -0,0 +1,268 @@
+        if val is not None:
+            self.validate_input(val)
+        self._input = val
+
+    def validate_input(self, input: Dict[str, Any]) -> None:
+        """Validates input params and contents/type/shape/keys of values.
+
+        Calls validate method on individual key/value pairs in input, and
+        handles validations requiring knowledge of other parameters.
+
+        Parameters
+        ----------
+        input : dict
+            Input file contents parsed to dict.
+
+        Raises
+        ------
+        ValueError, TypeError whenever an input parameter fails one of it's
+        value/type/shape/key validations.
+        """
+
+        self._validate_required_parameters(input)
+        if "data" in input.keys():
+            require_keys = "type" not in input["data"].keys()
+            require_keys |= "name" not in input["data"].keys()
+            if require_keys:
+                raise ValueError("Data 'type' and 'name' must not be empty.")
+            if input["data"]["type"] == "GA_object":
+                if "workspace" not in input.keys():
+                    msg = "Input file must contain a 'workspace' path "
+                    msg += "for data type 'GA_object'."
+                    raise ValueError(msg)
+            if input["data"]["type"] in ["ubc_grav", "ubc_mag"]:
+                if "data_file" not in input.keys():
+                    msg = "Input file must contain an 'input_file' path"
+                    msg += " for data types 'ubc_grav' and 'ubc_mag'."
+                    raise ValueError(msg)
+        elif "input_mesh" in input.keys():
+            if "save_to_geoh5" not in input.keys():
+                msg = "Input file must contain a 'save_to_geoh5' path if"
+                msg += " 'input_mesh' provided."
+                raise ValueError(msg)
+            if "input_mesh_file" not in input.keys():
+                msg = "Input file must contain a 'input_mesh_file' path if"
+                msg += " 'input_mesh' provided."
+                raise ValueError(msg)
+        elif "forward_only" in input.keys():
+            if input["forward_only"] == True:
+                if "reference_model" not in input.keys():
+                    msg = "A reference model/value must be provided for "
+                    msg += "forward modeling"
+                    raise ValueError(msg)
+        elif "reference_model" in input.keys():
+            if "value" in input["reference_model"].keys():
+                v = input["reference_model"]["value"]
+                v = np.r_[v]
+                if v.shape[0] not in [1, 3]:
+                    msg = "Start model needs to be a scalar or 3 component vector"
+                    raise ValueError(msg)
+        elif "save_to_geoh5" in input.keys():
+            if "out_group" not in input.keys():
+                msg = "Input file must contain a 'out_group' string if"
+                msg += "'save_to_geoh5' provided."
+                raise ValueError(msg)
```

Comment:
Could we condense all of the above with a list of required keys attached to each property?